File: src/retrocast/adapters/paroutes.py
from __future__ import annotations

import logging
from collections import defaultdict
from collections.abc import Iterator, Mapping
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field, ValidationError

from retrocast.adapters.base import AdaptMode, RawRouteEntry
from retrocast.adapters.errors import (
    adapter_cycle_error,
    adapter_node_type_error,
    adapter_schema_error,
    adapter_target_mismatch,
)
from retrocast.chem import canonicalize_smiles, get_inchi_key
from retrocast.exceptions import AdapterLogicError, ChemError, InvalidSmilesError
from retrocast.models.route import Molecule, Reaction, Route
from retrocast.models.task import Target
from retrocast.typing import ReactionSmilesStr, SmilesStr

logger = logging.getLogger(__name__)


# SECTION: Raw PaRoutes Schema


class PaRoutesReactionMetadata(BaseModel):
    id: str = Field(..., alias="ID")
    rsmi: str | None = None
    smiles: str | None = None
    reaction_hash: str | None = None
    ring_breaker: bool | None = Field(None, alias="RingBreaker")


class PaRoutesBaseNode(BaseModel):
    smiles: str
    children: list[PaRoutesNode] = Field(default_factory=list)


class PaRoutesMoleculeInput(PaRoutesBaseNode):
    type: Literal["mol"]
    in_stock: bool = False


class PaRoutesReactionInput(PaRoutesBaseNode):
    type: Literal["reaction"]
    metadata: PaRoutesReactionMetadata
    children: list[PaRoutesMoleculeInput] = Field(default_factory=list)


PaRoutesNode = Annotated[PaRoutesMoleculeInput | PaRoutesReactionInput, Field(discriminator="type")]


# SECTION: Condition Slot Annotations


class ConditionSlotParseStatistics(BaseModel):
    malformed_rsmi_count: int = 0
    uncanonicalizable_token_count: int = 0
    uncanonicalizable_tokens: defaultdict[str, int] = Field(default_factory=lambda: defaultdict(int))

    @property
    def distinct_uncanonicalizable_token_count(self) -> int:
        return len(self.uncanonicalizable_tokens)

    @property
    def top_uncanonicalizable_tokens(self) -> list[tuple[str, int]]:
        return sorted(self.uncanonicalizable_tokens.items(), key=lambda item: (-item[1], item[0]))[:5]


def _extract_condition_slot(
    rsmi: str | None,
    *,
    condition_slot_parse_statistics: ConditionSlotParseStatistics | None = None,
) -> str | None:
    if not rsmi:
        return None

    parts = rsmi.split(">")
    if len(parts) != 3:
        if condition_slot_parse_statistics is not None:
            condition_slot_parse_statistics.malformed_rsmi_count += 1
        return None

    condition_slot = parts[1].strip()
    return condition_slot or None


def _parse_condition_slot_smiles(
    condition_slot: str,
    *,
    condition_slot_parse_statistics: ConditionSlotParseStatistics | None = None,
) -> list[SmilesStr]:
    parsed_smiles: list[SmilesStr] = []
    for token in condition_slot.split("."):
        token = token.strip()
        if not token:
            continue
        try:
            parsed_smiles.append(canonicalize_smiles(token, remove_mapping=True))
        except ChemError:
            if condition_slot_parse_statistics is not None:
                condition_slot_parse_statistics.uncanonicalizable_token_count += 1
                condition_slot_parse_statistics.uncanonicalizable_tokens[token] += 1

    return sorted(parsed_smiles)


def _build_condition_slot_annotations(
    *,
    source_id: str,
    rsmi: str | None,
    ring_breaker: bool | None,
    condition_slot_parse_statistics: ConditionSlotParseStatistics | None = None,
) -> dict[str, Any]:
    annotations: dict[str, Any] = {"source_id": source_id}
    if ring_breaker is not None:
        annotations["ring_breaker"] = ring_breaker

    condition_slot = _extract_condition_slot(
        rsmi,
        condition_slot_parse_statistics=condition_slot_parse_statistics,
    )
    if condition_slot is not None:
        annotations["condition_slot"] = condition_slot
        condition_slot_smiles = _parse_condition_slot_smiles(
            condition_slot,
            condition_slot_parse_statistics=condition_slot_parse_statistics,
        )
        if condition_slot_smiles:
            annotations["condition_slot_smiles"] = condition_slot_smiles

    return annotations


# SECTION: Adapter


class PaRoutesAdapter:
    def iter_raw_routes(
        self,
        raw_payload: Any,
        *,
        source_key: str | None = None,
    ) -> Iterator[RawRouteEntry]:
        if not isinstance(raw_payload, Mapping):
            target_id = source_key or "<unknown>"
            raise adapter_schema_error("paroutes", target_id, "expected route root or mapping of target ids to routes")

        is_route_root = raw_payload.get("type") == "mol" or "smiles" in raw_payload or "children" in raw_payload
        if is_route_root:
            target_id = source_key or "<unknown>"
            yield RawRouteEntry(
                payload=self._validate_route_root(raw_payload, target_id=target_id),
                source_key=source_key,
                source_order=1,
            )
            return

        for source_order, (target_id, raw_route) in enumerate(raw_payload.items(), start=1):
            if not isinstance(target_id, str):
                raise adapter_schema_error("paroutes", source_key or "<unknown>", "target id keys must be strings")
            yield RawRouteEntry(
                payload=self._validate_route_root(raw_route, target_id=target_id),
                source_key=target_id,
                source_order=source_order,
            )

    def _validate_route_root(self, raw_route: Any, *, target_id: str) -> PaRoutesMoleculeInput:
        try:
            return PaRoutesMoleculeInput.model_validate(raw_route)
        except ValidationError as exc:
            raise adapter_schema_error("paroutes", target_id, "invalid molecule route root") from exc

    def cast(
        self,
        raw_route: Any,
        *,
        mode: AdaptMode = "strict",
        target: Target | None = None,
    ) -> Route:
        target_id = target.id if target is not None else "<unknown>"
        raw_route = self._validate_route_root(raw_route, target_id=target_id)
        patent_ids = self._get_patent_ids(raw_route, mode=mode)
        if not patent_ids:
            raise AdapterLogicError(
                f"PaRoutes route for target '{target_id}' does not contain a patent id",
                code="adapter.patent_id_missing",
                context={"adapter": "paroutes", "target_id": target_id},
            )
        if len(patent_ids) > 1:
            raise AdapterLogicError(
                f"PaRoutes route for target '{target_id}' contains reactions from multiple patents",
                code="adapter.multiple_patents",
                context={"adapter": "paroutes", "target_id": target_id, "patent_ids": sorted(patent_ids)},
            )

        route_target = self._build_molecule(raw_route, visited=set(), mode=mode)
        if route_target is None:
            raise AdapterLogicError(
                "PaRoutes target molecule was pruned",
                code="adapter.target_pruned",
                context={"adapter": "paroutes", "target_id": target_id},
            )

        if target is not None:
            expected_smiles = canonicalize_smiles(target.smiles)
            if route_target.smiles != expected_smiles:
                raise adapter_target_mismatch(
                    "paroutes",
                    target.id,
                    expected_smiles=expected_smiles,
                    actual_smiles=route_target.smiles,
                )

        return Route(target=route_target, annotations={"patent_id": next(iter(patent_ids))})

    def _get_patent_ids(
        self,
        node: PaRoutesMoleculeInput,
        *,
        mode: AdaptMode,
        visited: set[str] | None = None,
    ) -> set[str]:
        if visited is None:
            visited = set()
        try:
            canon_smiles = canonicalize_smiles(node.smiles)
        except InvalidSmilesError:
            if mode == "prune":
                return set()
            raise
        if canon_smiles in visited:
            raise adapter_cycle_error("paroutes", canon_smiles)

        patent_ids: set[str] = set()
        new_visited = visited | {canon_smiles}
        if node.children:
            reaction_node = self._require_reaction_node(node.children[0], role="molecule child")
            patent_id = reaction_node.metadata.id.split(";", 1)[0].strip()
            if not patent_id:
                raise AdapterLogicError(
                    "PaRoutes reaction metadata contains an empty patent id",
                    code="adapter.patent_id_missing",
                    context={"adapter": "paroutes", "source_id": reaction_node.metadata.id},
                )
            patent_ids.add(patent_id)
            for reactant_node in reaction_node.children:
                patent_ids.update(self._get_patent_ids(reactant_node, mode=mode, visited=new_visited))
        return patent_ids

    def _require_reaction_node(self, node: PaRoutesNode, *, role: str) -> PaRoutesReactionInput:
        if isinstance(node, PaRoutesReactionInput):
            return node
        raise adapter_node_type_error("paroutes", expected="reaction", actual=node.type, role=role)

    def _build_molecule(
        self,
        raw_mol_node: PaRoutesMoleculeInput,
        visited: set[SmilesStr],
        *,
        mode: AdaptMode,
    ) -> Molecule | None:
        try:
            canon_smiles = canonicalize_smiles(raw_mol_node.smiles)
        except InvalidSmilesError:
            if mode == "prune":
                return None
            raise

        if canon_smiles in visited:
            raise adapter_cycle_error("paroutes", canon_smiles)

        if raw_mol_node.in_stock or not raw_mol_node.children:
            return Molecule(smiles=canon_smiles, inchikey=get_inchi_key(canon_smiles))

        if len(raw_mol_node.children) > 1:
            logger.warning("molecule %s has multiple child reactions; only the first is used", canon_smiles)

        reaction_node = self._require_reaction_node(raw_mol_node.children[0], role="molecule child")

        reactants = []
        for reactant_node in reaction_node.children:
            reactant = self._build_molecule(reactant_node, visited | {canon_smiles}, mode=mode)
            if reactant is not None:
                reactants.append(reactant)
        if not reactants:
            if mode == "prune":
                return None
            raise AdapterLogicError(
                "PaRoutes reaction has no reactants",
                code="adapter.reaction_empty",
                context={"adapter": "paroutes", "smiles": canon_smiles},
            )

        reaction = Reaction(
            reactants=reactants,
            mapped_reaction_smiles=ReactionSmilesStr(reaction_node.metadata.rsmi)
            if reaction_node.metadata.rsmi
            else None,
            annotations=_build_condition_slot_annotations(
                source_id=reaction_node.metadata.id,
                rsmi=reaction_node.metadata.rsmi,
                ring_breaker=reaction_node.metadata.ring_breaker,
            ),
        )
        return Molecule(smiles=canon_smiles, inchikey=get_inchi_key(canon_smiles), product_of=reaction)


# SECTION: Diagnostics


def analyze_condition_slots(
    raw_route: Mapping[str, Any],
    *,
    stats: ConditionSlotParseStatistics,
) -> None:
    def visit(node: Mapping[str, Any]) -> None:
        children = node.get("children")
        if not isinstance(children, list):
            return
        for child in children:
            if not isinstance(child, Mapping):
                continue
            if child.get("type") == "reaction":
                metadata = child.get("metadata")
                if isinstance(metadata, Mapping) and isinstance(metadata.get("ID"), str):
                    _build_condition_slot_annotations(
                        source_id=metadata["ID"],
                        rsmi=metadata.get("rsmi") if isinstance(metadata.get("rsmi"), str) else None,
                        ring_breaker=metadata.get("RingBreaker")
                        if isinstance(metadata.get("RingBreaker"), bool)
                        else None,
                        condition_slot_parse_statistics=stats,
                    )
            visit(child)

    visit(raw_route)

File: packages/retrocast-rs/crates/retrocast-core/src/adapters/paroutes.rs
use std::collections::{BTreeMap, BTreeSet, HashSet};

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value, json};

use super::{Adapter, RawRouteEntry, common};
use crate::{
    chem,
    error::Result,
    model::{Molecule, Reaction, Route, Target},
    route::{AdaptMode, normalize_reactants},
    schema::{CanonicalSmiles, ReactionSmiles},
};

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct ConditionSlotParseStatistics {
    pub malformed_rsmi_count: usize,
    pub uncanonicalizable_token_count: usize,
    pub uncanonicalizable_tokens: BTreeMap<String, usize>,
}

impl ConditionSlotParseStatistics {
    pub fn distinct_uncanonicalizable_token_count(&self) -> usize {
        self.uncanonicalizable_tokens.len()
    }

    pub fn top_uncanonicalizable_tokens(&self) -> Vec<(&str, usize)> {
        let mut tokens: Vec<_> = self
            .uncanonicalizable_tokens
            .iter()
            .map(|(token, count)| (token.as_str(), *count))
            .collect();
        tokens.sort_by(|left, right| right.1.cmp(&left.1).then_with(|| left.0.cmp(right.0)));
        tokens.truncate(5);
        tokens
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct PaRoutesAdapter;

impl Adapter for PaRoutesAdapter {
    fn name(&self) -> &'static str {
        "paroutes"
    }

    fn entries(&self, payload: Value, source_key: Option<&str>) -> Result<Vec<RawRouteEntry>> {
        match payload {
            Value::Array(routes) => routes
                .into_iter()
                .enumerate()
                .map(|(index, route)| {
                    validate_molecule(&route, self.name(), "route root")?;
                    Ok(RawRouteEntry {
                        payload: route,
                        source_key: source_key.map(str::to_owned),
                        source_row_index: Some(index + 1),
                        source_record_id: None,
                        target_hint_id: None,
                        target_hint_smiles: None,
                        source_order: Some(index + 1),
                    })
                })
                .collect(),
            Value::Object(routes) if looks_like_route(&routes) => {
                let route = Value::Object(routes);
                validate_molecule(&route, self.name(), "route root")?;
                Ok(vec![RawRouteEntry {
                    payload: route,
                    source_key: source_key.map(str::to_owned),
                    source_row_index: None,
                    source_record_id: None,
                    target_hint_id: None,
                    target_hint_smiles: None,
                    source_order: Some(1),
                }])
            }
            Value::Object(routes) => routes
                .into_iter()
                .enumerate()
                .map(|(index, (target_id, route))| {
                    validate_molecule(&route, self.name(), "route root")?;
                    Ok(RawRouteEntry {
                        payload: route,
                        source_key: Some(target_id),
                        source_row_index: None,
                        source_record_id: None,
                        target_hint_id: None,
                        target_hint_smiles: None,
                        source_order: Some(index + 1),
                    })
                })
                .collect(),
            _ => Err(common::schema(
                self.name(),
                "expected route root, route list, or target route mapping",
            )),
        }
    }

    fn cast(&self, raw_route: Value, mode: AdaptMode, target: Option<&Target>) -> Result<Route> {
        validate_molecule(&raw_route, self.name(), "route root")?;
        let patent_ids = collect_patent_ids(&raw_route, mode, &HashSet::new())?;
        if patent_ids.is_empty() {
            return Err(common::logic(
                self.name(),
                "adapter.patent_id_missing",
                "route does not contain a patent id",
            ));
        }
        if patent_ids.len() > 1 {
            return Err(common::logic(
                self.name(),
                "adapter.multiple_patents",
                format!("route contains reactions from multiple patents: {patent_ids:?}"),
            ));
        }
        let target_molecule = build_molecule(&raw_route, mode, &HashSet::new())?
            .ok_or_else(|| common::target_pruned(self.name()))?;
        if let Some(target) = target {
            common::require_target_match(
                &target_molecule.smiles,
                &target.smiles,
                &target.id,
                self.name(),
            )?;
        }
        Ok(Route {
            target: target_molecule,
            annotations: Map::from_iter([(
                "patent_id".to_owned(),
                json!(patent_ids.first().expect("patent set is nonempty")),
            )]),
            schema_version: Default::default(),
        })
    }
}

fn looks_like_route(object: &Map<String, Value>) -> bool {
    object.get("type").and_then(Value::as_str) == Some("mol")
        || object.contains_key("smiles")
        || object.contains_key("children")
}

fn validate_molecule(node: &Value, adapter: &'static str, path: &str) -> Result<()> {
    let object = node
        .as_object()
        .ok_or_else(|| common::schema(adapter, format!("{path}: molecule must be an object")))?;
    if object.get("type").and_then(Value::as_str) != Some("mol") {
        return Err(common::schema(
            adapter,
            format!("{path}.type: expected 'mol'"),
        ));
    }
    if object.get("smiles").and_then(Value::as_str).is_none() {
        return Err(common::schema(
            adapter,
            format!("{path}.smiles: string field is required"),
        ));
    }
    let children = children(object, adapter, path)?;
    for (index, reaction) in children.iter().enumerate() {
        validate_reaction(reaction, adapter, &format!("{path}.children.{index}"))?;
    }
    Ok(())
}

fn validate_reaction(node: &Value, adapter: &'static str, path: &str) -> Result<()> {
    let object = node
        .as_object()
        .ok_or_else(|| common::schema(adapter, format!("{path}: reaction must be an object")))?;
    if object.get("type").and_then(Value::as_str) != Some("reaction") {
        return Err(common::logic(
            adapter,
            "adapter.node_type_invalid",
            format!("{path}: expected reaction node"),
        ));
    }
    if object.get("smiles").and_then(Value::as_str).is_none() {
        return Err(common::schema(
            adapter,
            format!("{path}.smiles: string field is required"),
        ));
    }
    let metadata = object
        .get("metadata")
        .and_then(Value::as_object)
        .ok_or_else(|| common::schema(adapter, format!("{path}.metadata: object is required")))?;
    if metadata.get("ID").and_then(Value::as_str).is_none() {
        return Err(common::schema(
            adapter,
            format!("{path}.metadata.ID: string field is required"),
        ));
    }
    for (index, molecule) in children(object, adapter, path)?.iter().enumerate() {
        validate_molecule(molecule, adapter, &format!("{path}.children.{index}"))?;
    }
    Ok(())
}

fn children<'a>(
    object: &'a Map<String, Value>,
    adapter: &'static str,
    path: &str,
) -> Result<&'a [Value]> {
    object
        .get("children")
        .map(|children| {
            children
                .as_array()
                .map(Vec::as_slice)
                .ok_or_else(|| common::schema(adapter, format!("{path}.children: expected list")))
        })
        .transpose()
        .map(|children| children.unwrap_or_default())
}

fn collect_patent_ids(
    molecule: &Value,
    mode: AdaptMode,
    visited: &HashSet<CanonicalSmiles>,
) -> Result<BTreeSet<String>> {
    let object = molecule.as_object().expect("validated molecule");
    let raw_smiles = object["smiles"].as_str().expect("validated smiles");
    let smiles = match chem::normalize(raw_smiles) {
        Ok((smiles, _)) => smiles,
        Err(_) if mode == AdaptMode::Prune => return Ok(BTreeSet::new()),
        Err(error) => return Err(error),
    };
    if visited.contains(&smiles) {
        return Err(common::logic(
            "paroutes",
            "adapter.cycle_detected",
            format!("cycle at {smiles:?}"),
        ));
    }
    let reactions = children(object, "paroutes", "molecule")?;
    let Some(reaction) = reactions.first() else {
        return Ok(BTreeSet::new());
    };
    let reaction = reaction.as_object().expect("validated reaction");
    let source_id = reaction["metadata"]["ID"]
        .as_str()
        .expect("validated patent ID");
    let patent_id = source_id.split(';').next().unwrap_or_default().trim();
    if patent_id.is_empty() {
        return Err(common::logic(
            "paroutes",
            "adapter.patent_id_missing",
            format!("empty patent id in source {source_id:?}"),
        ));
    }
    let mut patent_ids = BTreeSet::from([patent_id.to_owned()]);
    let mut next_visited = visited.clone();
    next_visited.insert(smiles);
    for reactant in children(reaction, "paroutes", "reaction")? {
        patent_ids.extend(collect_patent_ids(reactant, mode, &next_visited)?);
    }
    Ok(patent_ids)
}

fn build_molecule(
    molecule: &Value,
    mode: AdaptMode,
    visited: &HashSet<CanonicalSmiles>,
) -> Result<Option<Molecule>> {
    let object = molecule.as_object().expect("validated molecule");
    let raw_smiles = object["smiles"].as_str().expect("validated smiles");
    let (smiles, inchikey) = match chem::normalize(raw_smiles) {
        Ok(identity) => identity,
        Err(_) if mode == AdaptMode::Prune => return Ok(None),
        Err(error) => return Err(error),
    };
    if visited.contains(&smiles) {
        return Err(common::logic(
            "paroutes",
            "adapter.cycle_detected",
            format!("cycle at {smiles:?}"),
        ));
    }
    let reactions = children(object, "paroutes", "molecule")?;
    if object
        .get("in_stock")
        .and_then(Value::as_bool)
        .unwrap_or(false)
        || reactions.is_empty()
    {
        return Ok(Some(Molecule {
            smiles,
            inchikey,
            product_of: None,
            annotations: Map::new(),
        }));
    }
    let reaction = reactions[0].as_object().expect("validated reaction");
    let mut next_visited = visited.clone();
    next_visited.insert(smiles.clone());
    let mut reactants = Vec::new();
    for reactant in children(reaction, "paroutes", "reaction")? {
        if let Some(reactant) = build_molecule(reactant, mode, &next_visited)? {
            reactants.push(reactant);
        }
    }
    if reactants.is_empty() {
        return if mode == AdaptMode::Prune {
            Ok(None)
        } else {
            Err(common::logic(
                "paroutes",
                "adapter.reaction_empty",
                format!("reaction for {smiles:?} has no reactants"),
            ))
        };
    }
    normalize_reactants(&mut reactants);
    let metadata = reaction["metadata"]
        .as_object()
        .expect("validated metadata");
    let source_id = metadata["ID"].as_str().expect("validated source id");
    let rsmi = metadata.get("rsmi").and_then(Value::as_str);
    let mapped_reaction_smiles = rsmi.map(ReactionSmiles::try_from).transpose()?;
    Ok(Some(Molecule {
        smiles,
        inchikey,
        product_of: Some(Box::new(Reaction {
            reactants,
            mapped_reaction_smiles,
            template: None,
            reagents: None,
            solvents: None,
            annotations: condition_annotations(
                source_id,
                rsmi,
                metadata.get("RingBreaker").and_then(Value::as_bool),
                None,
            ),
        })),
        annotations: Map::new(),
    }))
}

fn condition_annotations(
    source_id: &str,
    rsmi: Option<&str>,
    ring_breaker: Option<bool>,
    mut statistics: Option<&mut ConditionSlotParseStatistics>,
) -> Map<String, Value> {
    let mut annotations = Map::from_iter([("source_id".to_owned(), json!(source_id))]);
    if let Some(ring_breaker) = ring_breaker {
        annotations.insert("ring_breaker".to_owned(), json!(ring_breaker));
    }
    let Some(rsmi) = rsmi else {
        return annotations;
    };
    let parts: Vec<_> = rsmi.split('>').collect();
    if parts.len() != 3 {
        if let Some(statistics) = statistics.as_deref_mut() {
            statistics.malformed_rsmi_count += 1;
        }
        return annotations;
    }
    let slot = parts[1].trim();
    if slot.is_empty() {
        return annotations;
    }
    annotations.insert("condition_slot".to_owned(), json!(slot));
    let mut parsed = Vec::new();
    for token in slot
        .split('.')
        .map(str::trim)
        .filter(|token| !token.is_empty())
    {
        match chem::normalize_unmapped(token) {
            Ok((smiles, _)) => parsed.push(smiles),
            Err(_) => {
                if let Some(statistics) = statistics.as_deref_mut() {
                    statistics.uncanonicalizable_token_count += 1;
                    *statistics
                        .uncanonicalizable_tokens
                        .entry(token.to_owned())
                        .or_default() += 1;
                }
            }
        }
    }
    parsed.sort();
    if !parsed.is_empty() {
        annotations.insert("condition_slot_smiles".to_owned(), json!(parsed));
    }
    annotations
}

pub fn analyze_condition_slots(route: &Value, statistics: &mut ConditionSlotParseStatistics) {
    fn visit(node: &Value, statistics: &mut ConditionSlotParseStatistics) {
        let Some(children) = node
            .as_object()
            .and_then(|object| object.get("children"))
            .and_then(Value::as_array)
        else {
            return;
        };
        for child in children {
            if child.get("type").and_then(Value::as_str) == Some("reaction") {
                if let Some(metadata) = child.get("metadata").and_then(Value::as_object) {
                    if let Some(source_id) = metadata.get("ID").and_then(Value::as_str) {
                        condition_annotations(
                            source_id,
                            metadata.get("rsmi").and_then(Value::as_str),
                            metadata.get("RingBreaker").and_then(Value::as_bool),
                            Some(statistics),
                        );
                    }
                }
            }
            visit(child, statistics);
        }
    }
    visit(route, statistics);
}

#[cfg(test)]
mod tests {
    use serde_json::{Value, json};

    use super::{ConditionSlotParseStatistics, PaRoutesAdapter, analyze_condition_slots};
    use crate::{adapters::Adapter, route::AdaptMode};

    fn route() -> Value {
        json!({
            "type": "mol", "smiles": "CCO", "children": [{
                "type": "reaction", "smiles": "CCO",
                "metadata": {"ID": "US123;1", "rsmi": "C>O.not-smiles>CCO", "RingBreaker": false},
                "children": [{"type": "mol", "smiles": "C", "in_stock": true}]
            }]
        })
    }

    #[test]
    fn preserves_patent_and_condition_annotations() {
        let route = PaRoutesAdapter
            .cast(route(), AdaptMode::Strict, None)
            .unwrap();
        assert_eq!(route.annotations["patent_id"], "US123");
        let reaction = route.target.product_of.unwrap();
        assert_eq!(reaction.annotations["condition_slot"], "O.not-smiles");
        assert_eq!(reaction.annotations["condition_slot_smiles"][0], "O");
    }

    #[test]
    fn condition_diagnostics_are_nonfatal() {
        let mut statistics = ConditionSlotParseStatistics::default();
        analyze_condition_slots(&route(), &mut statistics);
        assert_eq!(statistics.uncanonicalizable_token_count, 1);
        assert_eq!(
            statistics.top_uncanonicalizable_tokens(),
            vec![("not-smiles", 1)]
        );
    }

    #[test]
    fn rejects_mixed_patents() {
        let mut route = route();
        route["children"][0]["children"][0]["children"] = json!([{
            "type": "reaction", "smiles": "C", "metadata": {"ID": "OTHER;1"},
            "children": [{"type": "mol", "smiles": "[H][H]"}]
        }]);
        assert!(
            PaRoutesAdapter
                .cast(route, AdaptMode::Strict, None)
                .is_err()
        );
    }
}
